### ISSUE TITLE
Utilization reporting

### ISSUE THREAD

**Issue by Jkrzy** [18F/tock#980]
As an 18F tock user, I want to be able to see accurate and complete summaries of  billable hours tocked to the target number of billable hours for last week, the last month, and FYTD.

For this initial effort we're going to focus only on 18F use cases and specifically:

1. Displaying utilization for an individual on their details page
2. Displaying utilization for all staff by business unit 

Below describes our destination for aggregating utilization. We'll break this down into smaller pieces and work incrementally in this direction.

Considerations for 18F employees:

 - Employees enter or depart 18F at any time
 - Employees utilization targets will vary w/ time ((80% is default, engineering supervisors are 40%, for example))
 - Employees may move across 18F business units at any time
 - Employees `billable` status may change at any time

###  Timecards
To accurately aggregate utilization information across time periods we need to be able to include only timecards which "count" towards utilization.

A timecard should be included when:

- Employee was a `current employee`
- Employee was in 18F 
- Employee was a `Billable employee`

To achieve this we need to write more attributes to each Timecard when created:

- Employee's organization - `UserData.organization`
- Employee status - `UserData.current_employee`
- Billable status - `UserData.is_billable`
- 18F status - `is_18f_employee`
- 18F unit - `UserData.unit`

### Calculating Utilization

For a given timecard we calculate utilization as follows:
```python
# `total_hours_tocked_to_excluded_projects` is the sum of 
# hours tocked to projects with `exclude_from_billability == True` 
# e.g. Out of Office
expected_hours = ( 
  (total_hours_tocked - total_hours_tocked_to_excluded_projects) 
  * timecard.billable_expectation
)

utilization = (total_hours_tocked_to_billable_projects / expected_hours)
```

### Aggregate utilization at these intervals

A `ReportingPeriod` is considered complete if it's `end_date` is today or in the past.
Timecards will only be included if they meet the criteria specified above.

#### Last week
For a given user, the timecard set is defined as:

> The Timecard, if any, associated with the most recently completed `ReportingPeriod`

`ReportingPeriod.objects.filter(end_date__lte=today).get()`

#### Last month
For a given user, the timecard set is defined as:

> The Timecards, if any, associated with the 4 most recently completed `ReportingPeriod`

`ReportingPeriod.objects.filter(end_date__lte=today).order_by('-start_date')[:4]`

#### Fiscal year to date
For a given user, the timecard set is defined as:

> The Timecards, if any, associated with the all completed `ReportingPeriod`s since the beginning of the current fiscal year.

**Comment by tbaxter-18f** [comment]
> Staff move across TTS and can have varying utilization targets (80% is default, engineering supervisors are 40%, for example) and those targets will vary with time.

I think we know that this will lose reliability as folks move around. The more stable they are, the more it will be reliable. But I think if someone moves into a less-billable role they'll be over-target (good), and it's fairly rare for someone to move to a more-billable role, which would show them under-target (bad). Which is to say I think we can accept some inevitable wobbliness in service of providing more useful info.

> This is a big lift and will require changes across our models, views, and front end.

We may want to break it down into smaller pieces. We don't have to do it all at once. In particular, it might make sense to do the short week timeframe first (I think that's already merged, right?) then maybe last 90 days, then maybe YTD. However we do it, I think we can carve this into palatably bite-sized pieces.

> To accurately aggregate utilization information across time periods we need to be able to include only timecards which "count" towards utilization.

I think we can short-circuit a lot of the logic above by simply using `Billable employee` as the check. Because if they're not billable, for whatever reason, then nothing else matters, right?

> To achieve this we need to write more attributes to each Timecard when created:

Why do you feel we need to write these attributes to the timecard? We have them on the user object, right? I think I'm missing something here.

**Comment by saraheckert** [comment]
Thank you for this. Is it possible to also add a piece for controls for supervisors to review the reports or is that already encompassed in a separate issue?

**Comment by Jkrzy** [comment]
Consolidating @tbaxter-18f's thoughts above and conversation in Slack and updating issue now...

**Comment by tbaxter-18f** [comment]
@saraheckert to answer your question, we've already moved the reports into a location supervisors can see them, and that's just waiting on a release.

At that point, the data will be wonky, but once we can all see it, we can swarm on cleaning it up.

There's also another issue (#965) to update the business units in the reports. They're a bit out of date right now.

**Comment by saraheckert** [comment]
sounds great, thank you for the update

**Comment by tbaxter-18f** [comment]
Done
